File: llm_harness/clients/aa_stats.py
"""Artificial Analysis model stats client with cache + scoring output."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
import json
import math
import os
from pathlib import Path
from typing import Any

from dotenv import load_dotenv
import httpx
import polars as pl

load_dotenv()

MODELS_URL = "https://artificialanalysis.ai/api/v2/data/llms/models"
CACHE_PATH = Path(".cache/aa_models.json")
OUTPUT_PATH = Path(".cache/aa_output.json")
LOOKBACK_DAYS = 365
REQUEST_TIMEOUT_SECONDS = 30.0
DEFAULT_CACHE_TTL_SECONDS = 60 * 60 * 12
BENCHMARK_BIAS_KEYS = ("hle", "terminalbench_hard", "lcr", "ifbench", "scicode")
SCORE_WEIGHTS = {
    "intelligence": 0.3,
    "benchmark_bias": 0.3,
    "price": 0.2,
    "speed": 0.2,
}
PERCENTILE_COLUMNS = (
    ("overall_score", "overall_percentile"),
    ("intelligence_score", "intelligence_percentile"),
    ("speed_score", "speed_percentile"),
    ("price_score", "price_percentile"),
)


def _remove_ids(value: Any) -> Any:
    if isinstance(value, list):
        return [_remove_ids(item) for item in value]
    if isinstance(value, dict):
        return {key: _remove_ids(child) for key, child in value.items() if key != "id"}
    return value


def _as_finite_float(value: Any) -> float | None:
    if value is None:
        return None
    try:
        numeric = float(value)
    except (TypeError, ValueError):
        return None
    return numeric if math.isfinite(numeric) else None


def _is_positive_finite(value: Any) -> bool:
    numeric = _as_finite_float(value)
    return numeric is not None and numeric > 0


def _signed_log(value: Any, *, invert: bool = False) -> float | None:
    numeric = _as_finite_float(value)
    if numeric is None or numeric <= 0:
        return None
    return float(-math.log(numeric) if invert else math.log(numeric))


def _mean(values: list[float | None]) -> float | None:
    numbers = [value for value in values if value is not None and math.isfinite(value)]
    if not numbers:
        return None
    return float(sum(numbers) / len(numbers))


def _weighted_mean(pairs: list[tuple[float | None, float]]) -> float | None:
    valid_pairs = [(value, weight) for value, weight in pairs if value is not None and math.isfinite(value) and math.isfinite(weight) and weight > 0]
    if not valid_pairs:
        return None
    weighted_sum = sum(value * weight for value, weight in valid_pairs)
    weight_sum = sum(weight for _, weight in valid_pairs)
    if weight_sum == 0:
        return None
    return float(weighted_sum / weight_sum)


def _load_cache() -> dict[str, Any]:
    return json.loads(CACHE_PATH.read_text(encoding="utf-8"))


def _write_json(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(payload, indent=2), encoding="utf-8")


def _fetch_and_cache_models(api_key: str | None) -> dict[str, Any]:
    if not api_key:
        msg = "Missing ARTIFICIALANALYSIS_API_KEY for refresh. Set it or use existing cache."
        raise ValueError(msg)

    with httpx.Client(timeout=REQUEST_TIMEOUT_SECONDS) as client:
        response = client.get(MODELS_URL, headers={"x-api-key": api_key})
    response.raise_for_status()

    payload = response.json()
    cache_payload = {
        "fetched_at_epoch_seconds": int(datetime.now(UTC).timestamp()),
        "status_code": response.status_code,
        "models": [_remove_ids(model) for model in payload.get("data", [])],
    }
    _write_json(CACHE_PATH, cache_payload)
    return cache_payload


def get_aa_stats() -> dict[str, Any]:
    """Return filtered + scored AA model stats JSON and write `.cache/aa_output.json`."""
    api_key = os.getenv("ARTIFICIALANALYSIS_API_KEY")
    refresh_cache = os.getenv("AA_REFRESH") == "1"
    cache_ttl_seconds = int(os.getenv("AA_CACHE_TTL_SECONDS", str(DEFAULT_CACHE_TTL_SECONDS)))

    if refresh_cache:
        cache_payload = _fetch_and_cache_models(api_key)
    else:
        try:
            cache_payload = _load_cache()
            age_seconds = int(datetime.now(UTC).timestamp()) - int(cache_payload["fetched_at_epoch_seconds"])
            if age_seconds > cache_ttl_seconds:
                cache_payload = _fetch_and_cache_models(api_key)
        except Exception:
            cache_payload = _fetch_and_cache_models(api_key)

    cutoff_date = (datetime.now(UTC) - timedelta(days=LOOKBACK_DAYS)).date().isoformat()
    enriched_models_by_row_id: dict[int, dict[str, Any]] = {}
    scoring_rows: list[dict[str, float | int | None]] = []

    for row_id, model in enumerate(cache_payload.get("models", [])):
        release_date = str(model.get("release_date") or "")
        pricing = model.get("pricing") or {}
        evaluations = model.get("evaluations") or {}

        blended_price = _as_finite_float(pricing.get("price_1m_blended_3_to_1"))
        input_price = _as_finite_float(pricing.get("price_1m_input_tokens"))
        output_price = _as_finite_float(pricing.get("price_1m_output_tokens"))
        ttfa = _as_finite_float(model.get("median_time_to_first_answer_token"))
        tps = _as_finite_float(model.get("median_output_tokens_per_second"))

        if release_date < cutoff_date:
            continue
        if not all(_is_positive_finite(value) for value in [blended_price, input_price, output_price, ttfa, tps]):
            continue

        intelligence_index = _as_finite_float(evaluations.get("artificial_analysis_intelligence_index"))
        coding_index = _as_finite_float(evaluations.get("artificial_analysis_coding_index"))

        intelligence_score = None
        if intelligence_index is not None and coding_index is not None:
            intelligence_score = float((2.0 * intelligence_index) + coding_index)

        benchmark_values = [_as_finite_float(evaluations.get(key)) for key in BENCHMARK_BIAS_KEYS]
        benchmark_bias_score = _mean([value for value in benchmark_values if value is not None and value > 0])

        price_score = _signed_log(blended_price, invert=True)
        speed_score = _mean([_signed_log(ttfa, invert=True), _signed_log(tps)])
        overall_score = _weighted_mean(
            [
                (intelligence_score, SCORE_WEIGHTS["intelligence"]),
                (benchmark_bias_score, SCORE_WEIGHTS["benchmark_bias"]),
                (price_score, SCORE_WEIGHTS["price"]),
                (speed_score, SCORE_WEIGHTS["speed"]),
            ]
        )

        enriched_model = {
            **model,
            "scores": {
                "overall_score": overall_score,
                "intelligence_score": intelligence_score,
                "benchmark_bias_score": benchmark_bias_score,
                "price_score": price_score,
                "speed_score": speed_score,
            },
            "percentiles": {
                "overall_percentile": None,
                "intelligence_percentile": None,
                "speed_percentile": None,
                "price_percentile": None,
            },
        }
        enriched_models_by_row_id[row_id] = enriched_model
        scoring_rows.append(
            {
                "row_id": row_id,
                "overall_score": overall_score,
                "intelligence_score": intelligence_score,
                "speed_score": speed_score,
                "price_score": price_score,
            }
        )

    if scoring_rows:
        df = pl.DataFrame(scoring_rows).drop_nulls("overall_score")
        if df.height > 0:
            df = df.with_columns(
                [
                    (pl.col(score_column).rank("max") / pl.col(score_column).drop_nulls().count() * 100.0).alias(percentile_column)
                    for score_column, percentile_column in PERCENTILE_COLUMNS
                ]
            ).sort("overall_score", descending=True)
            ranked_models: list[dict[str, Any]] = []
            for row in df.iter_rows(named=True):
                model = enriched_models_by_row_id[int(row["row_id"])]
                model["percentiles"] = {
                    "overall_percentile": row["overall_percentile"],
                    "intelligence_percentile": row["intelligence_percentile"],
                    "speed_percentile": row["speed_percentile"],
                    "price_percentile": row["price_percentile"],
                }
                ranked_models.append(model)
            enriched_models = ranked_models
        else:
            enriched_models = []
    else:
        enriched_models = []

    output_payload = {
        "fetched_at_epoch_seconds": cache_payload["fetched_at_epoch_seconds"],
        "status_code": cache_payload["status_code"],
        "models": enriched_models,
    }
    _write_json(OUTPUT_PATH, output_payload)
    return output_payload

File: llm-harness-js/src/clients/aaStats.ts
import { add, mean, multiply } from "mathjs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { resolve } from "node:path";

const MODELS_URL = "https://artificialanalysis.ai/api/v2/data/llms/models";
const CACHE_PATH = resolve(".cache/aa_models.json");
const OUTPUT_PATH = resolve(".cache/aa_output.json");
const LOOKBACK_DAYS = 365;
const REQUEST_TIMEOUT_MS = 30_000;
const DEFAULT_CACHE_TTL_SECONDS = 60 * 60 * 24;
const BENCHMARK_KEYS = [
  "hle",
  "terminalbench_hard",
  "lcr",
  "ifbench",
  "scicode",
] as const;

const SCORE_WEIGHTS = {
  intelligence: 0.3,
  benchmark_bias: 0.3,
  price: 0.2,
  speed: 0.2,
} as const;

type NumberOrNull = number | null;
type ModelCreator = {
  name?: string;
  slug?: string;
};

type Evaluations = {
  artificial_analysis_intelligence_index?: number | null;
  artificial_analysis_coding_index?: number | null;
  hle?: number | null;
  terminalbench_hard?: number | null;
  lcr?: number | null;
  ifbench?: number | null;
  scicode?: number | null;
  [key: string]: unknown;
};

type Pricing = {
  price_1m_blended_3_to_1?: number | null;
  price_1m_input_tokens?: number | null;
  price_1m_output_tokens?: number | null;
  [key: string]: unknown;
};

type BaseModel = {
  name?: string;
  slug?: string;
  release_date?: string;
  model_creator?: ModelCreator;
  evaluations?: Evaluations;
  pricing?: Pricing;
  median_output_tokens_per_second?: number | null;
  median_time_to_first_token_seconds?: number | null;
  median_time_to_first_answer_token?: number | null;
  [key: string]: unknown;
};

type Scores = {
  overall_score: NumberOrNull;
  intelligence_score: NumberOrNull;
  benchmark_bias_score: NumberOrNull;
  price_score: NumberOrNull;
  speed_score: NumberOrNull;
};

type Percentiles = {
  overall_percentile: NumberOrNull;
  intelligence_percentile: NumberOrNull;
  speed_percentile: NumberOrNull;
  price_percentile: NumberOrNull;
};

type ScoredModel = BaseModel & { scores: Scores };

export type AaEnrichedModel = BaseModel & {
  scores: Scores;
  percentiles: Percentiles;
};

type CachePayload = {
  fetched_at_epoch_seconds: number;
  status_code: number;
  models: BaseModel[];
};

export type AaOutputPayload = {
  fetched_at_epoch_seconds: number;
  status_code: number;
  models: AaEnrichedModel[];
};

function finiteNumbers(values: unknown[]): number[] {
  return values
    .filter((value) => value != null)
    .map((value) => Number(value))
    .filter((value) => Number.isFinite(value));
}

function meanOfFinite(values: unknown[]): NumberOrNull {
  const numbers = finiteNumbers(values);
  if (numbers.length === 0) {
    return null;
  }
  return Number(mean(numbers));
}

function isPositiveFinite(value: unknown): boolean {
  const numericValue = Number(value);
  return Number.isFinite(numericValue) && numericValue > 0;
}

function signedLog(value: unknown, invert = false): NumberOrNull {
  if (!isPositiveFinite(value)) {
    return null;
  }
  const numericValue = Number(value);
  return invert ? -Math.log(numericValue) : Math.log(numericValue);
}

function weightedMean(
  pairs: Array<{ value: NumberOrNull; weight: number }>,
): NumberOrNull {
  const validPairs = pairs.filter(
    (pair) =>
      pair.value != null &&
      Number.isFinite(pair.value) &&
      Number.isFinite(pair.weight) &&
      pair.weight > 0,
  );
  if (validPairs.length === 0) {
    return null;
  }
  const weightedSum = validPairs.reduce(
    (sum, pair) => sum + (pair.value as number) * pair.weight,
    0,
  );
  const weightSum = validPairs.reduce((sum, pair) => sum + pair.weight, 0);
  if (weightSum === 0) {
    return null;
  }
  return weightedSum / weightSum;
}

function percentileRank(values: unknown[], value: unknown): NumberOrNull {
  if (value == null) {
    return null;
  }
  const numericValue = Number(value);
  if (!Number.isFinite(numericValue)) {
    return null;
  }
  const finiteValues = finiteNumbers(values);
  if (finiteValues.length === 0) {
    return null;
  }
  const lessOrEqualCount = finiteValues.filter(
    (item) => item <= numericValue,
  ).length;
  return (lessOrEqualCount / finiteValues.length) * 100;
}

function removeIds<T>(value: T): T {
  if (Array.isArray(value)) {
    return value.map((item) => removeIds(item)) as T;
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>)
        .filter(([key]) => key !== "id")
        .map(([key, child]) => [key, removeIds(child)]),
    ) as T;
  }
  return value;
}

function computeScores(filteredModels: BaseModel[]): ScoredModel[] {
  return filteredModels.map((model) => {
    const intelligence = Number(
      model.evaluations?.artificial_analysis_intelligence_index,
    );
    const coding = Number(model.evaluations?.artificial_analysis_coding_index);
    const blendedPrice = model.pricing?.price_1m_blended_3_to_1;
    const ttfa = model.median_time_to_first_answer_token;
    const tps = model.median_output_tokens_per_second;

    const intelligenceRaw =
      Number.isFinite(intelligence) && Number.isFinite(coding)
        ? Number(add(multiply(2, intelligence), coding))
        : null;
    const priceRaw = signedLog(blendedPrice, true);
    const ttfaRaw = signedLog(ttfa, true);
    const tpsRaw = signedLog(tps);
    const intelligenceScore = intelligenceRaw;
    const benchmarkBiasScore = meanOfFinite(
      BENCHMARK_KEYS.map((key) => {
        if (!isPositiveFinite(model.evaluations?.[key])) {
          return null;
        }
        return Number(model.evaluations?.[key]);
      }),
    );
    const priceScore = priceRaw;
    const speedScore = meanOfFinite([ttfaRaw, tpsRaw]);

    return {
      ...model,
      scores: {
        overall_score: weightedMean([
          { value: intelligenceScore, weight: SCORE_WEIGHTS.intelligence },
          { value: benchmarkBiasScore, weight: SCORE_WEIGHTS.benchmark_bias },
          { value: priceScore, weight: SCORE_WEIGHTS.price },
          { value: speedScore, weight: SCORE_WEIGHTS.speed },
        ]),
        intelligence_score: intelligenceScore,
        benchmark_bias_score: benchmarkBiasScore,
        price_score: priceScore,
        speed_score: speedScore,
      },
    };
  });
}

function rankAndEnrichModels(
  models: BaseModel[],
  cutoffDate: string,
): AaEnrichedModel[] {
  const filteredModels = models.filter((model) => {
    return (
      (model.release_date ?? "") >= cutoffDate &&
      isPositiveFinite(model.pricing?.price_1m_blended_3_to_1) &&
      isPositiveFinite(model.pricing?.price_1m_input_tokens) &&
      isPositiveFinite(model.pricing?.price_1m_output_tokens) &&
      isPositiveFinite(model.median_time_to_first_answer_token) &&
      isPositiveFinite(model.median_output_tokens_per_second)
    );
  });

  const scoredModels = computeScores(filteredModels);
  const ranked = scoredModels
    .filter((model) => Number.isFinite(model.scores.overall_score))
    .sort(
      (left, right) =>
        (right.scores.overall_score ?? Number.NEGATIVE_INFINITY) -
        (left.scores.overall_score ?? Number.NEGATIVE_INFINITY),
    );

  const overallValues = ranked.map((model) => model.scores.overall_score);
  const intelligenceValues = ranked.map(
    (model) => model.scores.intelligence_score,
  );
  const speedValues = ranked.map((model) => model.scores.speed_score);
  const priceValues = ranked.map((model) => model.scores.price_score);

  return ranked.map((model) => ({
    ...model,
    percentiles: {
      overall_percentile: percentileRank(
        overallValues,
        model.scores.overall_score,
      ),
      intelligence_percentile: percentileRank(
        intelligenceValues,
        model.scores.intelligence_score,
      ),
      speed_percentile: percentileRank(speedValues, model.scores.speed_score),
      price_percentile: percentileRank(priceValues, model.scores.price_score),
    },
  }));
}

async function loadCache(): Promise<CachePayload> {
  const content = await readFile(CACHE_PATH, "utf-8");
  return JSON.parse(content) as CachePayload;
}

async function writeJson(path: string, payload: unknown): Promise<void> {
  await mkdir(resolve(".cache"), { recursive: true });
  await writeFile(path, JSON.stringify(payload, null, 2), "utf-8");
}

async function fetchAndCacheModels(
  apiKey: string | undefined,
): Promise<CachePayload> {
  if (!apiKey) {
    throw new Error(
      "Missing ARTIFICIALANALYSIS_API_KEY for refresh. Set it or use existing cache.",
    );
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
  const response = await fetch(MODELS_URL, {
    headers: { "x-api-key": apiKey },
    signal: controller.signal,
  });
  clearTimeout(timeout);

  if (!response.ok) {
    throw new Error(`Artificial Analysis request failed: ${response.status}`);
  }

  const payload = (await response.json()) as { data: BaseModel[] };
  const cachePayload: CachePayload = {
    fetched_at_epoch_seconds: Math.floor(Date.now() / 1000),
    status_code: response.status,
    models: payload.data.map((model) => removeIds(model)),
  };
  await writeJson(CACHE_PATH, cachePayload);
  return cachePayload;
}

export async function getAAStats(): Promise<AaOutputPayload> {
  const apiKey = process.env.ARTIFICIALANALYSIS_API_KEY;
  const refreshCache = process.env.AA_REFRESH === "1";
  const cacheTtlSeconds = Number(
    process.env.AA_CACHE_TTL_SECONDS ?? DEFAULT_CACHE_TTL_SECONDS,
  );

  let cachePayload: CachePayload;
  if (!refreshCache) {
    try {
      const cached = await loadCache();
      const ageSeconds =
        Math.floor(Date.now() / 1000) - cached.fetched_at_epoch_seconds;
      if (ageSeconds <= cacheTtlSeconds) {
        cachePayload = cached;
      } else {
        cachePayload = await fetchAndCacheModels(apiKey);
      }
    } catch {
      cachePayload = await fetchAndCacheModels(apiKey);
    }
  } else {
    cachePayload = await fetchAndCacheModels(apiKey);
  }

  const cutoffDate = new Date(Date.now() - LOOKBACK_DAYS * 24 * 60 * 60 * 1000)
    .toISOString()
    .slice(0, 10);

  const outputPayload: AaOutputPayload = {
    fetched_at_epoch_seconds: cachePayload.fetched_at_epoch_seconds,
    status_code: cachePayload.status_code,
    models: rankAndEnrichModels(cachePayload.models, cutoffDate),
  };

  await writeJson(OUTPUT_PATH, outputPayload);
  return outputPayload;
}
